### ISSUE TITLE
Feature request: reverse proxy subfolder (URL Base) option

### ISSUE THREAD

**Issue by whitestrake** [meeb/tubesync#255]
Hi there - I'd just like to make a request, if it's possible, for a configurable URL Base so that I can serve the app from behind my HTTPS proxy in a subfolder, e.g. `example.com/tubesync/`, alongside others such as `example.com/sonarr/` and `example.com/radarr/` etc; those others are good examples of this configurable functionality.

No problems at all serving it on its own domain, of course, just a request for neatness and personal preference. Thanks!

**Comment by meeb** [comment]
I believe you can set this via the `DJANGO_FORCE_SCRIPT_NAME` environment variable. Try setting it to `/tubesync` and see if that works behind your reverse proxy. This looks like a duplicate of #18 ?

**Comment by whitestrake** [comment]
Ahh, jeez, I searched "reverse proxy" and "subdirectory" but didn't search "subfolder" even though I put it in the title of this issue... apologies. I'll give it a shot and let you know if it works. Thanks!

**Comment by whitestrake** [comment]
Hmm, doesn't look like I've had a win, here.

```
whitestrake in opt/docker/mediaserver at omnius is 🐳 v20.10.17
➜ grep -A10 tubesync: docker-compose.yml
  tubesync:
    image: ghcr.io/meeb/tubesync:latest
    restart: unless-stopped
    environment:
      TZ: "Snip"
      PUID: "1000"
      PGID: "1000"
      DJANGO_FORCE_SCRIPT_NAME: "/tubesync"
    ports:
      - 4848:4848/tcp
    volumes:
      - ./tubesync:/config
      - /mnt/cadmus/media/youtube:/downloads

whitestrake in opt/docker/mediaserver at omnius is 🐳 v20.10.17
➜ docker compose exec tubesync env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=2e5efd37b059
TERM=xterm
TZ=snip
PUID=1000
PGID=1000
DJANGO_FORCE_SCRIPT_NAME=/tubesync
DEBIAN_FRONTEND=noninteractive
HOME=/root
LANGUAGE=en_US.UTF-8
LANG=en_US.UTF-8
LC_ALL=en_US.UTF-8
PYTHONPATH=/app:

whitestrake in opt/docker/mediaserver at omnius is 🐳 v20.10.17
➜ curl -IL localhost:4848/tubesync
HTTP/1.1 404 Not Found
Server: nginx
Date: Fri, 22 Jul 2022 10:40:41 GMT
Content-Type: text/html; charset=utf-8
Content-Length: 186
Connection: keep-alive
Vary: Accept-Encoding
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
Referrer-Policy: same-origin


whitestrake in opt/docker/mediaserver at omnius is 🐳 v20.10.17
➜ curl -IL localhost:4848
HTTP/1.1 200 OK
Server: nginx
Date: Fri, 22 Jul 2022 10:40:44 GMT
Content-Type: text/html; charset=utf-8
Content-Length: 16100
Connection: keep-alive
Vary: Accept-Encoding
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
Referrer-Policy: same-origin
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
```

**Comment by meeb** [comment]
You'll still need to forward `https://reverseproxy/tubesync` to `dockerhost:4848`. All `FORCE_SCRIPT_NAME` does is set the `SCRIPT_NAME` for requests. See https://docs.djangoproject.com/en/4.0/ref/settings/#force-script-name - this won't magically move TubeSync to listen on a different URI prefix in its own container (that would also require rewriting parts of the nginx config bundled with TubeSync), it just makes it behave behind an upstream reverse proxy on a different URI prefix.

**Comment by whitestrake** [comment]
> this won't magically move TubeSync to listen on a different URI prefix in its own container

To be clear, this is the behavior I was expecting from URL Base functionality. That would mean that the reverse proxy does not have to alter the URI it receives before sending it on to the backend.

i.e. I request `example.com/tubesync/`, reach my proxy, and my proxy requests `tubesync:4848/tubesync/`.

---

That said, instead of reverse proxying `example.com/tubesync/` to `tubesync:4848/tubesync/` I have configured my reverse proxy (Caddy) to strip the URI prefix before proxying (so requests to `/tubesync/` are reverse proxied to `/` instead as in your example).

This results in a 200 OK response for `example.com/tubesync/` but the page then makes a subrequest for `/tubesync.css` (rather than `/tubesync/tubesync.css`) which then fails because I'm not reverse proxying the webroot to the app, only `/tubesync`. Instead of the app I see a blank white page rendered.

I note that this is the exact behavior I expect of stripping the URI prefix without taking additional steps to ensure anchor links are corrected (such as would be the case with Sonarr/Radarr etc which correctly reference JS/CSS resources with the supplied URL Base, or other apps which reference those resources _relatively_ instead of absolute to the webroot and therefore function the same regardless of which subfolder you reverse proxy from).

**Comment by meeb** [comment]
The static media loaded by TubeSync is handled via Django so this should update the resource URLs to match the `SCRIPT_NAME` prefix. I'll check the source of the SASS compiler to see if this is respecting the WSGI prefix, it should be and last time I tested this (admittedly well over a year ago) it did work as expected, all be it behind nginx as a reverse proxy and not Caddy.

**Comment by whitestrake** [comment]
Gotcha. Just to clarify, with `DJANGO_FORCE_SCRIPT_NAME` set (as above demonstrated by running `env` in the container), when I `curl localhost:4848` the tag in the HTML output remains `<link type="text/css" rel="stylesheet" href="/static/styles/tubesync.css" media="screen,projection"/>`.

**Comment by meeb** [comment]
Yeah looks like I might have to also set `STATIC_URL` to work as well to get the SASS tag to behave. I'll need to find some time to spin up a test instance with a suitable reverse proxy to confirm it.

**Comment by whitestrake** [comment]
No problem!

Just on the off chance you want to test on Caddy, you can use the following Caddyfile:

```
http://:8080
handle /tubesync*
redir /tubesync /tubesync/
uri strip_prefix /tubesync
reverse_proxy localhost:4848
```

Downloading the single static binary from https://caddyserver.com/download and running with `./caddy run -config /path/to/Caddyfile`.

That configuration should reverse proxy to `localhost:4848` (change as appropriate), passing requests from `http://localhost:8080/tubesync/` back to `/` upstream with much the same configuration I'm currently attempting to use.

**Comment by whitestrake** [comment]
Just to update this - I've actually found that it _does_ work... kind of!

Hyperlinks within the generated page are correctly referenced to the provided `DJANGO_FORCE_SCRIPT_NAME`.

That means that when you set this but still access it via `localhost:4848`, it works but all the links point to `/tubesync/*`. So it's partially doing something. The static CSS asset, though, is the part that _doesn't_ get referenced correctly, and so the first page load through the proxy doesn't actually show me anything to click on anyway.

**Comment by meeb** [comment]
Yep that would be the `STATIC_ROOT` being incorrect, but the `SCRIPT_NAME` being set correctly.

**Comment by whitestrake** [comment]
Can I set `STATIC_ROOT` a la `DJANGO_FORCE_STATIC_ROOT` or similar?

**Comment by meeb** [comment]
No not easily yourself, until I push a patch for it.

**Comment by meeb** [comment]
Try that (after waiting ~30 mins for the container image to build). Usage would be something like (in ENV vars):

`DJANGO_FORCE_SCRIPT_NAME=/tubesync/`
`DJANGO_STATIC_URL=/tubesync/static/`

**Comment by whitestrake** [comment]
OK, new problem.

With `DJANGO_STATIC_URL=/tubesync/static` in the container, the CSS link is now issued for `/tubesync/static/styles/tubesync.css` (excellent!).

However... it is serving the CSS locally from `localhost:4848/tubesync/static/styles/tubesync.css`. This is actually _not_ a good thing for the reverse proxy, which is currently stripping the `/tubesync` prefix before sending it upstream because this is required to resolve the main index.

I could resolve this by _not_ stripping the URI prefix for the `/tubesync/static` path, but, in order to match the behaviour of the index, it should issue the link for `/tubesync/static` but it should _respond_ to `/static`.

**Comment by meeb** [comment]
Given you are using your own reverse proxy you can try bypassing TubeSync's built in nginx reverse proxy. Try setting these:

`LISTEN_HOST=0.0.0.0`
`LISTEN_PORT=8080`

This exposes the internal gunicorn server which nginx talks to. You can then forward port 4848 to 8080 on your container / composer side. Then plop Caddy in front of port 4848. This might work, untested though.

**Comment by whitestrake** [comment]
I'll give it a shot. Is that in addition to the two env vars already set (for script name and static URL?)

**Comment by meeb** [comment]
Yes, in addition to the other ENV vars.

**Comment by whitestrake** [comment]
Tested, same issue;

`localhost:8080` prefixes all links to `/tubesync`, but while the index listens on `/`, the static asset listens only to `/tubesync`, meaning the reverse proxy must be configured separately for static assets compared to the rest of the application.

**Comment by meeb** [comment]
I've had a few mins to actually test this now rather than just offer blind suggestions. There's no obvious native support to handle this properly in Django, luckily you can patch in the path in the WSGI environ. I've added a single new ENV var to set, you can remove the other 4 now, usage:

`DJANGO_URL_PREFIX=/tubesync/`

It must end with a `/` or you'll get a generic application error and some errors in the logs.

Wait for the container to build with the above patch and see if that resolves it for you.

**Comment by whitestrake** [comment]
OK, I've got the index working now and it serves CSS correctly, no problem!

It still serves the index on the web root `/` but now it _also_ serves the index in the specified subfolder `/tubesync/`. That means the last bit I needed to do was _disable_ my reverse proxy's URI prefix stripping. Now the reverse proxy is from `/tubesync/*` to `/tubesync/*` and the main page works.

...however, the _subfolders_ still don't respond under the `/tubesync` directory.

```
whitestrake in opt/docker/mediaserver at omnius is 🐳 v20.10.17
➜ dc exec tubesync env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=04c373827037
TERM=xterm
PGID=1000
DJANGO_URL_PREFIX=/tubesync/
TZ=[snip]
PUID=1000
DEBIAN_FRONTEND=noninteractive
HOME=/root
LANGUAGE=en_US.UTF-8
LANG=en_US.UTF-8
LC_ALL=en_US.UTF-8
PYTHONPATH=/app:

whitestrake in opt/docker/mediaserver at omnius is 🐳 v20.10.17
➜ curl -IL localhost:4848/tubesync/
HTTP/1.1 200 OK
Server: nginx
Date: Mon, 25 Jul 2022 04:50:42 GMT
Content-Type: text/html; charset=utf-8
Content-Length: 16131
Connection: keep-alive
Vary: Accept-Encoding
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
Referrer-Policy: same-origin
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block


whitestrake in opt/docker/mediaserver at omnius is 🐳 v20.10.17
➜ curl -IL localhost:4848/tubesync/sources
HTTP/1.1 404 Not Found
Server: nginx
Date: Mon, 25 Jul 2022 04:50:55 GMT
Content-Type: text/html; charset=utf-8
Content-Length: 186
Connection: keep-alive
Vary: Accept-Encoding
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
Referrer-Policy: same-origin


whitestrake in opt/docker/mediaserver at omnius is 🐳 v20.10.17
➜ curl -IL localhost:4848/sources
HTTP/1.1 200 OK
Server: nginx
Date: Mon, 25 Jul 2022 04:51:01 GMT
Content-Type: text/html; charset=utf-8
Content-Length: 6998
Connection: keep-alive
Vary: Accept-Encoding
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
Referrer-Policy: same-origin
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
```

**Comment by meeb** [comment]
OK! Seems some WSGI environment variables may or may not have a `/` prefixed for some reason. Try the above update. Note this won't require URI stripping in Caddy.

**Comment by meeb** [comment]
Did that last patch actually fix this for you?

**Comment by whitestrake** [comment]
Yes, I've pulled the latest and given it a shot. It seems to be working 100%; all subpages, resources, and links loading as expected, no issues encountered over the last day. 👍 

My working Docker Compose implementation with the latest version is:

```
  tubesync:
    image: ghcr.io/meeb/tubesync:latest
    restart: unless-stopped
    environment:
      TZ: "[snip]"
      PUID: "1000"
      PGID: "1000"
      DJANGO_URL_PREFIX: "/tubesync/"
    volumes:
      - ./tubesync:/config
      - /mnt/cadmus/media/youtube:/downloads
    labels:
      caddy: "example.com"
      caddy.handle: "/tubesync*"
      caddy.handle.import: "secure"
      caddy.handle.redir: "/tubesync /tubesync/"
      caddy.handle.reverse_proxy: "{{upstreams 4848}}"
      com.centurylinklabs.watchtower.enable: "true"
```

For reference, the Caddy labels produce configuration equivalent to a Caddyfile like:

```
example.com {
  handle /tubesync* {
    import secure # imports a snippet from elsewhere implementing SSO
    redir /tubesync /tubesync/
    reverse_proxy tubesync:4848
  }
}
```

(Although in my case there are many more other containers merging Caddy config for other app reverse proxies into this site block, into other directories and the webroot itself, that Tubesync is now playing nicely with.)

Thank you very much @meeb for the app and for rapidly iterating to get this feature request working! Cheers!

**Comment by meeb** [comment]
No problem, got there in the end. I'm sure it'll be useful for others as a feature.

**Comment by Majoraslayer** [comment]
I know this is an older issue, but I'm having trouble adapting this to work with Nginx.  I'm trying to access the app externally at mydomain.exa/d02 (for example) but can only get a blank page with TubeSync in the title bar.  Any chance I could get some help feeding this with the right options in Nginx, or let me know what I'm missing in my environment variables?  To build on this existing feature request, it would be great to get a usage guide on setting up reverse proxies for the app as a reference in the main README.

My Docker run config:

```
docker run -d \
  --name=tubesync \
  -e PUID=1000 \
  -e GUID=1000 \
  -e TZ=America/Chicago \
  -e HTTP_USER=myusername \
  -e HTTP_PASS=mypassword \
  -e DJANGO_URL_PREFIX="/d02/" \
  -p 4848:4848 \
  -v /path/to/docker/configs/tubesync:/config \
  -v /mnt/DownloadsHDD/YouTube:/downloads \
  --restart unless-stopped \
  ghcr.io/meeb/tubesync:latest
```

My Nginx config:

```
server {
        listen         443 ssl http2;
        server_name    mydomain.exa;
        location /d02/ {
          proxy_pass http://192.168.1.40:4848/;
          proxy_set_header Host $host;
          proxy_set_header X-Real-IP $remote_addr;
          proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
          proxy_set_header X-Forwarded-Proto $scheme;
        }
       # My Certbot certificate stuff here
}
```

**Comment by meeb** [comment]
@Majoraslayer the tubesync container itself comes bundled with nginx, if you're using nginx as your upstream you might have better luck just bypassing the container nginx entirely. Set `LISTEN_HOST` to `0.0.0.0` then point your nginx server to the gunicorn WSGI server on port 8080 directly, `proxy_pass http://tubesync:8080;` or whatever the host / IP would be for your container stack. Double-chaining nginx reverse proxies is possible, just annoying.

**Comment by Majoraslayer** [comment]
Fortunately I did manage to get it working without bypassing the internal Nginx instance, though that would probably have been the better approach.  I edited my original comment to mention what I did in case it helps the next person trying to figure it out.  Nginx has a slightly different approach to reverse proxy than Caddy, so all I had to do is was include the base path in the proxy_pass directive as well.  Thanks for following up with me though!
